src/triage/component/catwalk/predictors.py

```diff
@@ -150,83 +140,24 @@ def _write_to_db(
-                        [
-                            prediction.model_id,
-                            prediction.entity_id,
-                            prediction.as_of_date,
-                            prediction.score,
-                            prediction.label_value,
-                            prediction.rank_abs,
-                            prediction.rank_pct,
-                            prediction.matrix_uuid,
-                            prediction.test_label_timespan,
-                        ]
-                    )
-                f.seek(0)
-                postgres_copy.copy_from(f, Prediction_obj, self.db_engine, format="csv")
-        else:
-            logging.info(
-                "as_of_date not found as part of matrix index, using "
-                "matrix metadata end_time as as_of_date"
-            )
-            temp_df = pandas.DataFrame({"score": predictions})
-            rankings_abs = temp_df["score"].rank(method="dense", ascending=False)
-            rankings_pct = temp_df["score"].rank(
-                method="dense", ascending=False, pct=True
-            )
-            for entity_id, score, label, rank_abs, rank_pct in zip(
-                matrix_store.index,
-                predictions,
-                labels,
-                rankings_abs,
-                rankings_pct,
+        def generate_prediction_objects():
+            for index, score, label in zip(
+                matrix_store.index, predictions, labels
             ):
-                db_objects.append(
-                    Prediction_obj(
-                        model_id=int(model_id),
-                        entity_id=int(entity_id),
-                        as_of_date=matrix_store.metadata["end_time"],
-                        score=round(float(score), 10),
-                        label_value=int(label) if not math.isnan(label) else None,
-                        rank_abs=int(rank_abs),
-                        rank_pct=round(float(rank_pct), 10),
-                        matrix_uuid=matrix_store.uuid,
-                        test_label_timespan=test_label_timespan,
-                        **misc_db_parameters
-                    )
+                entity_id, as_of_date = index
+                yield Prediction_obj(
+                    model_id=int(model_id),
+                    entity_id=int(entity_id),
+                    as_of_date=as_of_date,
+                    score=float(score),
+                    label_value=int(label) if not math.isnan(label) else None,
+                    matrix_uuid=matrix_store.uuid,
+                    test_label_timespan=test_label_timespan,
+                    **misc_db_parameters
                 )
-
-            try:
-                session = self.sessionmaker()
-                session.bulk_save_objects(db_objects)
-                session.commit()
-            finally:
-                session.close()
+        save_db_objects(self.db_engine, generate_prediction_objects())
```

Comment:
This could also just be a generator expression. (I prefer them inline, but it could even be assigned, anything like: `prediction_stream = (…)`.)

```suggestion
        save_db_objects(self.db_engine, (
            Prediction_obj(
                model_id=int(model_id),
                entity_id=int(entity_id),
                as_of_date=as_of_date,
                score=float(score),
                label_value=int(label) if not math.isnan(label) else None,
                matrix_uuid=matrix_store.uuid,
                test_label_timespan=test_label_timespan,
                **misc_db_parameters
            )
            for ((entity_id, as_of_date), score, label) in zip(
                matrix_store.index, predictions, labels
            )
        ))
```